### DISCUSSION TITLE
Revising Dependabot config and version specifiers of dev dependencies

### DISCUSSION THREAD

**Discussion by sisp** [copier-org/copier#1242] [Ideas]
I've stumbled over Dependabot's current behavior in this project and wonder whether it should be revised.

As far as I can tell (it's difficult to verify), Dependabot's behavior given our quite minimal configuration

https://github.com/copier-org/copier/blob/a5da81cec8e240341acf64c6634fdb54d2d0db14/.github/dependabot.yml#L8-L11

(with regard to the relevant settings) is this:

* [`allow`](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#allow) defaults to `direct`. This means, all (runtime and dev) dependencies listed in `pyproject.toml` are checked for new versions, but indirect/transitive dependencies are ignored.
* [`versioning-strategy`](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#versioning-strategy) defaults to `lockfile-only`. This means, only `poetry.lock` is updated and `pyproject.toml` is never updated.

The combination of those two settings means, all (runtime and dev) dependencies listed in `pyproject.toml` are checked and their versions are updated _only in `poetry.lock`_ when new versions are available. It also means, none of the indirect/transitive dependencies are updated (unless Dependabot detects security vulnerabilities), so while the direct dependencies remain up-to-date, (typically) the majority of dependencies in the dependency tree remains _outdated_. I fail to see the rationale for such behavior.

Also, I wonder about the motivation for using a version specifier other than `==` (or equivalently, none), i.e. exact pinning, _for all dev dependencies_ because dev dependencies are always installed from the lock file, i.e. pinned versions, so specifying lower bounds has little meaning except it increases the search space of the dependency resolver when the minimum version isn't the most recent version. And for code quality tools like formatters or linters, which may introduce, e.g., new rules even in non-major releases resulting in different formatting/linting output, pinning in combination with explicit version updates (assisted by Dependabot), which require/incur changes in the code to satisfy, e.g., new rules, seems much more sensible to me than using lower bounds and having the `poetry.lock` versions diverge from the minimum versions in `pyproject.toml`.

To summarize, I wonder whether the following changes would be sensible:

* Use pinned versions _for dev dependencies_ in `pyproject.toml`.

* Update `.github/dependabot.yml`

    * to update _all_ (i.e. also indirect/transitive) dependencies in `poetry.lock`, and
    * to update _pinned dev dependencies_ in `pyproject.toml`

    which might look like this (untested):

    ```diff
     - package-ecosystem: "pip"
       directory: "/"
       schedule:
         interval: "daily"
    +  allow:
    +    - dependency-type: all
    +  versioning-strategy: lockfile-only
    +
    +- package-ecosystem: "pip"
    +  directory: "/"
    +  schedule:
    +    interval: "daily"
    +  allow:
    +    - dependency-type: development
    +  versioning-strategy: increase
    ```

    Here, I assume that `versioning-strategy: increase` increases also pinned versions and also to new major versions. If pinning in `pyproject.toml` blocks `versioning-strategy: increase` from doing anything, then we could resort to using lower bounds.

**Comment by sisp** [comment]
Odd, specifying `package-ecosystem: "pip"` twice in this way seems to be disallowed, giving the following error:

```
The property '#/updates/2' is a duplicate. Update configs must have a unique combination of 'package-ecosystem', 'directory', and 'target-branch'
```

So it seems to be impossible to have different versioning strategies for dependency types. :thinking: Why should this not be allowed? :shrug:

**Reply by sisp** [reply]
Seems related to https://github.com/dependabot/dependabot-core/issues/1778. I'm shocked that Dependabot does not support this. :scream: Perhaps Renovate Bot is worth looking into? It supports [lock file maintenance](https://docs.renovatebot.com/configuration-options/#lockfilemaintenance):

> You can use `lockFileMaintenance` to refresh lock files to keep them up-to-date.
>
> When Renovate performs `lockFileMaintenance` it deletes the lock file and runs the relevant package manager. That package manager creates a new lock file, where all dependency versions are updated to the latest version. Renovate then commits that lock file to the update branch and creates the lock file update PR.

**Comment by yajo** [comment]
I agree to update all, and not only direct dependencies. However that will mean having more Dependabot PRs. Not really a big problem, but it would be nice to automerge if green somehow. 

I'd like to keep the current strategy. It seems correct to me to update just the lock file if possible, but to update the manifest if needed. That's happening in https://github.com/copier-org/copier/pull/1243 for example.

Lower bounds in dev dependencies are not really used, but also not harmful. We could leave them as they are, or just use `*`.

**Reply by sisp** [reply]
I believe I could live with using `*` as the version specifier for dev dependencies. The only downside, IMO, is the reduced transparency which version of a dev dependency is actually being used when not specifying a version pin. But Dependabot seems to be incapable of updating the pins in the manifest _only for dev dependencies_ and updating the lock file for _all_ (i.e. also indirect) dependencies.

So the change your suggesting is this, right?

* Update the version specifiers of all dev dependencies to `*`.

* Update the Dependabot config:

    ```diff
     - package-ecosystem: "pip"
       directory: "/"
       schedule:
         interval: "daily"
    +  allow:
    +    - dependency-type: all
    ```

    So we'd keep the versioning strategy but extend the lock file updates also to indirect dependencies, if I read the Dependabot docs right.

Regarding the amount of Dependabot PRs, I think it might be nice to keep individual PRs for updates of direct dependencies but have a single PR for a all indirect dependencies. But I don't think Dependabot supports this either. Even if we auto-merged Dependabot PRs upon passing CI checks, those commits would pollute the Git history quite a bit. 🤔

**Reply by yajo** [reply]
Having atomic updates is better IMHO. For example, sometimes I need to tweak something for the nix build, and it's easier that way.

BTW what's the downside of having lower bounds in dev deps? 🤔

**Reply by sisp** [reply]
> BTW what's the downside of having lower bounds in dev deps? 🤔

I think the result is typically fine because the versions are locked and typically only increased over time, but the lower bound is often arbitrary (e.g. the latest version at the time when the dependency got added) and may become wrong (e.g. when using a more recent feature without raising the lower bound to the version in which it got introduced, or e.g. when using a formatter that formats code differently in a recent version than in an older version and you wouldn't notice because you're on the newer version now). So, my point is rather about theoretical than practical correctness, yet it bothers me. :laughing:

**Reply by yajo** [reply]
In general terms I'm not much in favor of changing things just because yes. If something isn't broken, I prefer to not fix it.

Anyways I don't wanna be too dogmatic about it, so if you prefer we can put those asterisks in the dependencies. 

The other improvement actually adds value so it's more important IMHO.

**Comment by sisp** [comment]
@yajo Are you aware that Renovate Bot supports [updating Nix dependencies](https://docs.renovatebot.com/modules/manager/nix/)? [Dependabot doesn't](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#package-ecosystem). Just thinking whether checking out Renovate Bot might be worth some time, also regarding the original topic, but also because of Nix support.

**Reply by yajo** [reply]
It'd love that!

I actually asked for support on dependabot some time ago: https://github.com/dependabot/dependabot-core/issues/7340, but it seems stale.
